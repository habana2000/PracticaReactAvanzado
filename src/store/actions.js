import { areAdvertsLoaded, getAdvert } from './selectors';
import {
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_REQUEST,
  UI_RESET_ERROR,
  ADVERT_CREATED_SUCCESS,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_FAILURE,
} from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});

export const authLogin =
  (credentials, rememberPassword) =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(authLoginRequest());
    try {
      await service.auth.login(credentials, rememberPassword);
      // Logged in
      dispatch(authLoginSuccess());
      // Redirect to pathname
      const to = router.state.location.state?.from?.pathname || '/';
      router.navigate(to);
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };

export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT,
});

export const authLogout =
  () =>
  async (dispatch, _getState, { service }) => {
    await service.auth.logout();
    dispatch(authLogoutSuccess());
  };

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertsLoaded =
  () =>
  async (dispatch, getState, { service }) => {
    if (areAdvertsLoaded(getState())) {
      return;
    }
    dispatch(advertsLoadedRequest());
    try {
      const adverts = await service.adverts.getLatestAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = advert => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedFailure = error => ({
  type: ADVERT_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertLoad =
  advertId =>
  async (dispatch, getState, { service }) => {
    // debugger;
    const isLoaded = getAdvert(advertId)(getState());
    if (isLoaded) {
      return;
    }
    dispatch(advertLoadedRequest());
    try {
      const advert = await service.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };

export const advertCreateRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreateSuccess = advert => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
});

export const advertCreateFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  error: true,
  payload: error,
});

export const advertCreate =
  advert =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(advertCreateRequest());
    try {
      const { id } = await service.adverts.createAdvert(advert);
      const createdAdvert = await service.adverts.getAdvert(id);
      dispatch(advertCreateSuccess(createdAdvert));
      router.navigate(`/adverts/${id}`);
      return createdAdvert;
    } catch (error) {
      dispatch(advertCreateFailure(error));
    }
  };

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
