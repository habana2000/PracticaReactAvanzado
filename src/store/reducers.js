import {
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERT_LOADED_SUCCESS,
  UI_RESET_ERROR,
} from './types';

export const defaultState = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  tags: {
    areLoaded: false,
    data: []
  },
  ui: {
    isLoading: false,
    error: null,
  },
};


export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  if (action.type === ADVERT_LOADED_SUCCESS) {
    return { ...state, data: [action.payload] };
  }
  if (action.type === ADVERT_CREATED_SUCCESS) {
    return { ...state, data: [action.payload].concat(state.data) };
  }

  return state;
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { isLoading: false, error: action.payload };
  }

  if (/_REQUEST$/.test(action.type)) {
    return { isLoading: true, error: null };
  }

  if (/_SUCCESS$/.test(action.type)) {
    return { isLoading: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  return state;
}


