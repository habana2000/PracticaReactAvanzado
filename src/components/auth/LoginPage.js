import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Button from '../shared/Button';
import FormField from '../shared/FormField';

import './LoginPage.css';
import { authLogin, uiResetError } from '../../store/actions';
import { getUi } from '../../store/selectors';

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
  });
  const [rememberPassword, setRememberPassword] = useState(true);

  const resetError = () => {
    dispatch(uiResetError());
  };

  const handleSubmit = async event => {
    event.preventDefault();
    dispatch(authLogin(credentials, rememberPassword));
  };

  const handleChange = event => {
    setCredentials({
      ...credentials,
      [event.target.name]: event.target.value,
    });
  };

  const handleCheckboxChange = event => {
    setRememberPassword(rememberPassword ? false : true);
  };

  const buttonDisabled =
    isLoading || !credentials.email || !credentials.password;

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to Adverts page</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="phone, email or username"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.email}
          autofocus
        />
        <FormField
          type="password"
          name="password"
          label="password"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.password}
        />
        <label>
        <input
          type="checkbox"
          name="rememberPassword"
          value= {rememberPassword? rememberPassword : false}
          checked={rememberPassword === true}
          onChange={handleCheckboxChange}
          // onClick={handleCheckboxChange}
          />
          Remember password
        </label>
        <Button
          type="submit"
          variant="primary"
          className="loginForm-submit"
          disabled={buttonDisabled}
        >
          Log in
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
