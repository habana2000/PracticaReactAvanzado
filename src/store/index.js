import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';

import * as service from '../service';

import * as reducers from './reducers';
import * as actionCreators from './actions';

const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({
  features: {
    pause: true, // start/pause recording of dispatched actions
    lock: true, // lock/unlock dispatching actions and side effects
    persist: true, // persist states on page reloading
    export: true, // export history of actions in a file
    import: 'custom', // import history of actions from a file
    jump: true, // jump back and forth (time travelling)
    skip: true, // skip (cancel) actions
    reorder: true, // drag and drop actions in the history list
    dispatch: true, // dispatch custom actions or action creators
    test: true, // generate tests for the selected actions
  },  
  actionCreators,
});


const logger = store => next => action => {
  if (action.type) {
    console.group(action.type);
    console.info('dispatching', action, store.getState());
    const result = next(action);
    console.log('next state', store.getState());
    console.groupEnd();
    return result;
  }
  return next(action);
};

const timestamp = () => next => action => {
  return next({
    ...action,
    meta: {
      ...action.meta,
      timestamp: Date.now(),
    },
  });
};

const failureRedirects = (router, redirectsMap) => () => next => action => {
  const result = next(action);

  if (action.error) {
    const redirect = redirectsMap[action.payload.status];
    if (redirect) {
      router.navigate(redirect);
    }
  }
  return result;
};

export default function configureStore(preloadedState, { router }) {
  const middleware = [
    thunk.withExtraArgument({ service, router }),
    timestamp,
    failureRedirects(router, { 401: '/login', 404: '/404' }),
    logger,
  ];

  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware)),
  );
  return store;
}
