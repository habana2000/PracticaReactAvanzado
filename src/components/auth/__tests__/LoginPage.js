import { render, screen } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers';
import { authLogin, uiResetError } from '../../../store/actions';
import userEvent from '@testing-library/user-event';

jest.mock('../../../store/actions');

describe('LoginPage', () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = defaultState;
        state.ui.error = error;
        
        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
      <LoginPage />
      </Provider>,
      );
    };
    
    test('snapshot', () => {
      const { container } = renderComponent();
      expect(container).toMatchSnapshot();
    });

    /*
        
    test('should dispatch authLogin action', () => {
      const email = 'xroca@vilamatica.com';
    const password = '1234';

    renderComponent();

    const usernameInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();

    // fireEvent.change(usernameInput, { target: { value: email } });
    // fireEvent.change(passwordInput, { target: { value: password } });
    userEvent.type(usernameInput, email);
    userEvent.type(passwordInput, password);

    expect(submitButton).toBeEnabled();

    // fireEvent.click(submitButton);
    userEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email, password });
  });

  test('should display an error', () => {
    const error = { message: 'Unauthorized' };
    renderComponent(error);

    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();

    // fireEvent.click(errorElement);
    userEvent.click(errorElement);

    expect(uiResetError).toHaveBeenCalled();
  });

  */
});
