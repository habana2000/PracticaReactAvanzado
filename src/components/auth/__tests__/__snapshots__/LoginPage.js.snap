// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Adverts page
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            phone, email or username
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <label>
        <input
          checked=""
          name="rememberPassword"
          type="checkbox"
          value="true"
        />
        Remember password
      </label>
      <button
        class="sc-beqWaB ctOVvj loginForm-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
